### PR TITLE
Add a RESTEasy Multipart extension

### LINKED ISSUE #880: Introduce a RESTEasy Multipart extension
RESTEasy has rich support for the `multipart/*` and `multipart/form-data` mime types. The multipart mime format is used to pass lists of content bodies. Multiple content bodies are embedded in one message. `multipart/form-data` is often found in web application HTML Form documents and is generally used to upload files. The form-data format is the same as other multipart formats, except that each inlined piece of content has a name associated with it.

It should be supported in Protean too.

### LINKED ISSUE #10323: UTF-8 encoding problem with MultiPart/RestEasy
**Describe the bug**
I develop a rest API with MULTIPART_FORM. But I'm having problems with @Consumes charset. I sent some params to the controller which includes German characters. And it didn't work as expected. The characters are not encoded properly, what can also be seen in debugging the properties in VSCode.

Some of the characters that cause the problem:
* ä, Ä, ü, Ü, ö, Ö, ß

Instead, I receive:
* ��

**Expected behavior**
* Receive all Umlaut, etc. UTF-8

**To Reproduce**
Using httpie:
```
http -f POST localhost:8080/general content="Test-Ä" file@testFile.png
```

Response:
```
HTTP/1.1 200 OK
Content-Length: 11
Content-Type: text/plain;charset=UTF-8

Test-��
```

Ressource:
```
import org.jboss.resteasy.annotations.providers.multipart.MultipartForm;
import javax.ws.rs.Consumes;
import javax.ws.rs.POST;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;

public class FeedbackResource {

    @POST
    @Path("/general")
    @Produces(MediaType.TEXT_PLAIN)
    @Consumes(MediaType.MULTIPART_FORM_DATA+";charset=UTF-8")
    public String postForm(@MultipartForm final FeedbackBody feedback) {
        return feedback.content;
    }
}
```

Model:
```
package org.acme;

import org.jboss.resteasy.annotations.providers.multipart.PartType;

import javax.ws.rs.FormParam;
import javax.ws.rs.core.MediaType;

public class FeedbackBody {

    private byte[] file;
    public byte[] getFile() {
        return file;
    }

    @FormParam("file")
    @PartType(MediaType.APPLICATION_OCTET_STREAM)
    public void setFile(byte[] file) {
        this.file = file;
    }

    @FormParam("fileName")
    @PartType(MediaType.TEXT_PLAIN)
    public String fileName;

    @FormParam("content")
    @PartType(MediaType.TEXT_PLAIN+";charset=UTF-8")
    public String content;
}
```

**Configuration**
```properties
# Add your application.properties here, if applicable.
<properties>
    <compiler-plugin.version>3.8.1</compiler-plugin.version>
    <maven.compiler.parameters>true</maven.compiler.parameters>
    <maven.compiler.source>1.8</maven.compiler.source>
    <maven.compiler.target>1.8</maven.compiler.target>
    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
    <quarkus-plugin.version>1.5.1.Final</quarkus-plugin.version>
    <quarkus.platform.artifact-id>quarkus-universe-bom</quarkus.platform.artifact-id>
    <quarkus.platform.group-id>io.quarkus</quarkus.platform.group-id>
    <quarkus.platform.version>1.5.1.Final</quarkus.platform.version>
    <surefire-plugin.version>2.22.1</surefire-plugin.version>
  </properties>
```

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: 
Darwin Mac.local 19.5.0 Darwin Kernel Version 19.5.0: Tue May 26 20:41:44 PDT 2020; root:xnu-6153.121.2~2/RELEASE_X86_64 x86_64
 - Output of `java -version`:
openjdk version "11.0.6" 2020-01-14
OpenJDK Runtime Environment GraalVM CE 19.3.1 (build 11.0.6+9-jvmci-19.3-b07)
OpenJDK 64-Bit Server VM GraalVM CE 19.3.1 (build 11.0.6+9-jvmci-19.3-b07, mixed mode, sharing)
 - Quarkus version or git rev:
1.5.1.Final
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): 
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: /Users/.m2/wrapper/dists/apache-maven-3.6.3-bin/1iopthnavndlasol9gbrbg6bf2/apache-maven-3.6.3
Java version: 11.0.6, vendor: Oracle Corporation, runtime: /Users/christian/.sdkman/candidates/java/19.3.1.r11-grl
Default locale: en_GB, platform encoding: UTF-8
OS name: "mac os x", version: "10.15.5", arch: "x86_64", family: "mac"

### PR DESCRIPTION
**Pull request by gsmet**
Fixes https://github.com/quarkusio/quarkus/issues/10323 by adding a new `quarkus-resteasy-multipart`.
Fixes #880

When using the extension, the default charset is UTF-8 instead of US-ASCII as it looks like a much saner default. It's configurable via Quarkus configuration properties.

### CODE REVIEW

**Comment by gsmet** [comment]
@evialle @ejba I tried to implement an extension to fix #10323 in this PR.

But I can't get the test to pass, even if I'm positive the filter is taken into account.

If someone has any idea of what I'm doing wrong.

/cc @geoand

**Comment by geoand** [comment]
I'll pass it on to @gastaldi for the time being. Tag me in if needed :)

**Comment by gsmet** [comment]
@evialle if you could put together a small reproducer showing that your filter solves the issue, that would be helpful. Maybe I'm missing something obvious.

**Comment by gsmet** [comment]
@asoldano could you have a look at the principle of that one? I'm changing the default for `input-part` default charset in Quarkus as we had a lot of people complaining about it.

In 2020, I think the default should be UTF-8.

**Comment by gsmet** [comment]
@geoand interested in your feedback on this one.

**Comment by geoand** [comment]
I just have one question:

Would it make sense to not have a separate extension, but simply enable the filter conditionally at build time if the multipart dependency is present?

**Comment by gastaldi** [comment]
> Would it make sense to not have a separate extension, but simply enable the filter conditionally at build time if the multipart dependency is present?

That makes sense, as long as the necessary changes are restricted to just a filter setting :+1:

**Comment by gsmet** [comment]
It’s a very common usage. To be honest I wanted an extension for that one since a few months. 

Having an extension exposes it and I think it’s the right thing to do. 

The nice perk is that it won’t break existing applications not using the extension.

**Comment by gastaldi** [comment]
Funny you mention that, I had the same impression in https://github.com/quarkusio/quarkus/pull/879 :wink:

**Comment by gsmet** [comment]
Ah ah good one, you got me here :). I think I totally underestimated back in the time the number of people using multipart. 

We have a good reason to do it now so I would say let’s do it.

**Comment by geoand** [comment]
This time @gastaldi gets to say: "I told you so" :laughing:

**Comment by geoand** [comment]
I'm +1 for this

File: extensions/resteasy-multipart/runtime/src/main/java/io/quarkus/resteasy/multipart/runtime/ResteasyMultipartRuntimeConfig.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    @ConfigGroup
+    public static class InputPartConfigGroup {
+
+        /**
+         * Default charset.
+         * <p>
+         * Note that the default value is UTF-8 which is different from RESTEasy's default value US-ASCII.
+         */
+        @ConfigItem(defaultValue = "UTF-8")
+        public Charset defaultCharset;
```
Comment (gastaldi) [inline extensions/resteasy-multipart/runtime/src/main/java/io/quarkus/resteasy/multipart/runtime/ResteasyMultipartRuntimeConfig.java:28]: I know strong typing is important but why a `java.nio.charset.Charset` if you only use the name?

Comment (gsmet) [reply extensions/resteasy-multipart/runtime/src/main/java/io/quarkus/resteasy/multipart/runtime/ResteasyMultipartRuntimeConfig.java:28]: Mostly to be consistent with the other places. And this way it gets somehow validated by the converter.

File: extensions/resteasy-multipart/runtime/src/main/java/io/quarkus/resteasy/multipart/runtime/ResteasyMultipartRuntimeConfig.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
+package io.quarkus.resteasy.multipart.runtime;
+
+import java.nio.charset.Charset;
+
+import io.quarkus.runtime.annotations.ConfigGroup;
+import io.quarkus.runtime.annotations.ConfigItem;
+import io.quarkus.runtime.annotations.ConfigPhase;
+import io.quarkus.runtime.annotations.ConfigRoot;
+
+@ConfigRoot(name = "resteasy.multipart", phase = ConfigPhase.RUN_TIME)
```
[the remark this answers is not in the document]
Comment (gsmet) [reply extensions/resteasy-multipart/runtime/src/main/java/io/quarkus/resteasy/multipart/runtime/ResteasyMultipartRuntimeConfig.java:10]: Hmmm, isn't the bean instantiated at runtime? So you could change the value with a `-D...` in another environment for instance.

**Comment by gsmet** [comment]
I think this is ready and I answered all questions?

@gastaldi can I get your approval if you're good with it?

**Review by gastaldi — APPROVED** [review]
LGTM, good job!

**Comment by gsmet** [comment]
Thanks!

**Pull request merged**